src/Build/Instance/TaskRegistry.cs

```diff
@@ -1215,18 +1210,13 @@ internal ParameterGroupAndTaskElementRecord ParameterGroupAndTaskBody
             /// <returns>true if the task can be created by the factory, false if it cannot be created</returns>
             internal bool CanTaskBeCreatedByFactory(string taskName, string taskProjectFile, IDictionary<string, string> taskIdentityParameters, TargetLoggingContext targetLoggingContext, ElementLocation elementLocation)
             {
-                // Keep a cache of task identities which have been checked against the factory, this is useful because we ask this question everytime we get a registered task record or a taskFactory wrapper.
-                if (_taskNamesCreatableByFactory == null)
-                {
-                    _taskNamesCreatableByFactory = new Dictionary<RegisteredTaskIdentity, object>(RegisteredTaskIdentity.RegisteredTaskIdentityComparer.Exact);
-                }
-
                 RegisteredTaskIdentity taskIdentity = new RegisteredTaskIdentity(taskName, taskIdentityParameters);
 
                 // See if the task name as already been checked against the factory, return the value if it has
                 object creatableByFactory = null;
-                if (!_taskNamesCreatableByFactory.TryGetValue(taskIdentity, out creatableByFactory))
+                if (!_taskNamesCreatableByFactory.Value.TryGetValue(taskIdentity, out creatableByFactory))
                 {
+                    // Multiple threads can enter here for a single taskIdentity and create the factory, only one will be stored in the dictionary
```

Comment:
The `finally` block here is a bit of catch (haha - unitended pun) - as currently in case of unhandled exception an explicit null will be inserted into the dictionary (in finally); with `GetOrAdd` this is not possible.
While I *think* such behavior was not fully intended, I do not understand all the possible corner cases - so didn't want to change this

Reply:
Why is it not possible to add null with `GetOrAdd`? I believe the `Func<TKey,TValue>` can legally return null and have it added as a value.

Reply:
That is possible.
What is not possible is to add a null and throw an exception (if GetOrAdd throws, then value is not added and vice versa). The previous code was adding the value in `finally` - so in case of exception it'd add null and exception would still be thrown.

Reply:
I see! Ok, there may be ways around it (e.g. switch from `finally` to `catch` and do a `TryAdd(..., null)` there, but the perf benefit of calling only on operation on the happy path is probably negligible. Looks good!